groovy-mode.el

```diff
@@ -329,7 +312,77 @@ The function name is the second group in the regexp.")
+               (not (groovy--comment-p matched))
+               (let* ((line-beg (- (match-beginning 0) (line-beginning-position)))
+                      (cur-line (groovy--current-line))
+                      (str (buffer-substring-no-properties (line-beginning-position) (match-beginning 0)))
+                      (match-str (s-trim (match-string 0)))
+                      (open-char (match-string 1)))
+                 (message "match-str: %s" match-str)
+                 (save-match-data
+                   (unless (string-match (rx bol (* space) (eval match-str) (* space) eol) cur-line)
+                     ;; if ends in '=' (and not '==' or '=~') then it's a var assignment, highlight
+                     (if (s-ends-with-p "=" match-str)
+                         (let ((next-char (char-to-string (char-after (match-end 0)))))
+                           (and (not (equal next-char "="))
+                                (not (equal next-char "~"))
+                                ;; make sure this isn't in an annotation
+                                (not (string-match
+                                      (rx "@" (+ alphanumeric)
+                                          (* space)
+                                          (zero-or-one (seq "(" (* (not (any ")")))))
+                                          eol)
+                                      str))))
+                       ;; otherwise if the string doesn't end in '=' it could still be an uninitialized var
+                       ;; so we check for def or type
+                       (and
+                        ;; if the preceding non-whitespace char is '=' then it's not a var
+                        (not (string-match (rx "=" (* space) eol) str))
+                        ;; make sure we're not in a list: [a, b, c]
+                        (not (string-match (rx "[" (* (not (any "]"))) eol) str))
+                        ;; if not, look for declarations at line beginning
+                        (or (and (string-match groovy-declaration-regex str)
+                                 ;; if declaration make sure there's no = between the type and the var
+                                 (not (string-match "=" (substring cur-line (match-end 0) line-beg))))
+                            ;; or it's a list like '(a, b, c) = list'
+                            (let ((start
+                                   (if (equal open-char "(")
+                                       line-beg
+                                     (when (string-match (rx "(" (+ (not (any ")"))) eol) str)
+                                       (match-end 0)))))
+                              (or (not start)
+                                  (string-match (rx bol (+ (not (any ")"))) ")" (* space) (or "=" "{"))
+                                                (substring cur-line start nil)))))))))))
+          ;; keep searching
+          (groovy--variable-names-search limit)))))
```

Comment:
Heya! Thanks for the feedback!

The keyword funcs are passed a limit that (from what I've seen) is just outside the buffer's viewable area, so usually until you scroll no more than what can fit on the page is parsed and highlighted. I looked at some other modes and this was the most common way to implement it.